### PR TITLE
fix(mcp): return tool execution failures as isError results

### LINKED ISSUE #254: [Bug]: MCP tool execution failures are returned as JSON-RPC errors instead of isError results
## Is there an existing issue for the same bug?

- [x] I have checked the existing issues.

## Environment

Current main branch. Both RPC dispatch paths in `memoria/crates/memoria-mcp/src/server.rs` are affected (stdio and the owned dispatcher used by HTTP), across all three tool modules: `tools.rs`, `git_tools.rs` and `remote.rs`.

## Background

MCP draws a deliberate distinction between two kinds of `tools/call` failure:

- **Protocol errors** — the tool does not exist, the request is malformed. These are JSON-RPC errors. The host application handles them; the model never sees them.
- **Tool execution errors** — the tool ran but the operation failed (validation rejected the input, storage was unavailable, a merge conflicted). These are returned as a *successful* JSON-RPC result whose payload carries `isError: true` and the error text in `content`.

The distinction exists so the model can see what went wrong and correct itself. An execution error delivered as a JSON-RPC error is, for most hosts, invisible to the model: the host surfaces it as a tool-transport failure rather than feeding the text back into the conversation. The agent then has no idea why its call failed and cannot adapt.

## Actual behavior

Memoria collapses both categories into JSON-RPC error `-32000`.

In `memoria/crates/memoria-mcp/src/server.rs`, every `tools/call` outcome is mapped through the same closure in both dispatchers (lines 322 and 376):

    let internal_err = |e: anyhow::Error| McpRpcError {
        code: -32000,
        message: e.to_string(),
    };

`tools::call`, `git_tools::call` and the remote client all return `anyhow::Error` for any failure, so a rejected argument and an unreachable database are indistinguishable at the wire level. The unknown-tool case is folded in too — `tools.rs:1301` returns `Err(anyhow!("Unknown tool: {name}"))`, which also becomes `-32000` rather than an invalid-params error.

The `isError` field is never produced anywhere in the crate. Each module has its own success wrapper (`tools.rs:1314`, `git_tools.rs:2434`, `remote.rs:52`), all identical and all unconditional:

    fn mcp_text(text: &str) -> Value {
        json!({"content": [{"type": "text", "text": text}]})
    }

There is no corresponding error wrapper.

## Expected behavior

- Tool *execution* failures return a successful JSON-RPC result of the form:

      {"content": [{"type": "text", "text": "<error description>"}], "isError": true}

- Protocol-level failures remain JSON-RPC errors, with the unknown-tool case using an invalid-params code rather than the generic `-32000`.
- The error text returned to the model is actionable — it should say what was rejected and what a valid call looks like, since the model's only chance to correct itself is that string.
- Execution errors surfaced through `isError` are not recorded as `rpc_success = 0` in `mem_api_call_log`, since they are successful protocol exchanges; protocol errors still are.

## Steps to reproduce

1. Start the Memoria MCP server and complete the `initialize` handshake.
2. Issue a `tools/call` that fails inside the tool — for example `memory_store` with empty `content`, which `parse_store_content` rejects (`tools.rs:107`).
3. Observe the response is a JSON-RPC error:

       {"jsonrpc":"2.0","id":2,"error":{"code":-32000,"message":"content is required"}}

   Expected:

       {"jsonrpc":"2.0","id":2,"result":{"content":[{"type":"text","text":"content is required"}],"isError":true}}

4. Observe in a host such as Claude Desktop or Cursor that the model is not shown the `content is required` text and cannot retry with a corrected argument.

## Additional information

Observed in production against the currently deployed build, across all user databases, from 2026-08-27 to 2026-09-18 — 10 `-32000` responses:

| tool | count |
|---|---|
| `memory_store` | 4 |
| `memory_feedback` | 2 |
| `memory_retrieve` | 1 |
| `memory_merge` | 1 |
| `memory_diff` | 1 |
| `memory_apply` | 1 |

The volume is low, but each occurrence is a user's agent failing a memory operation with no recoverable signal. `memory_store` leading the list is the worst case: the memory is silently not written, and the agent continues believing it was.

Note that #213 (`memory_capabilities` returns plain text instead of structured JSON) and #211 / #216 (`memory_store` accepts invalid input without validation) touch the same result-shaping layer. Tightening validation as those issues ask for will *increase* the number of execution errors, so it is worth fixing this issue first or alongside them — otherwise better validation makes the agent experience worse rather than better.

MCP `tools/call` error-handling specification (2024-11-05, the version Memoria currently advertises):
https://modelcontextprotocol.io/specification/2024-11-05/server/tools

## Acceptance criteria

- Add an error wrapper alongside `mcp_text` in `tools.rs`, `git_tools.rs` and `remote.rs` that emits `content` plus `isError: true`.
- Separate execution failures from protocol failures in both dispatch paths in `server.rs`; return the former as results and keep the latter as JSON-RPC errors.
- Return an invalid-params error for unknown tool names instead of `-32000`.
- Stop counting `isError` results as `rpc_success = 0` in the call log.
- Cover both dispatch paths with tests asserting that a failing tool call yields a result with `isError: true` and a non-empty `content` message, and that an unknown tool still yields a JSON-RPC error.

### PR DESCRIPTION
**Pull request by loveRhythm1990**
## What type of PR is this?

- [x] fix (bug fix)

## Which issue(s) this PR fixes

Fixes #254

## What this PR does / why we need it

Tool execution failures currently become JSON-RPC errors or unmarked success text, making it difficult for clients to present actionable failures to the model. Return execution failures as tool results with `isError: true`, while keeping malformed tool-call envelopes and unknown tools as JSON-RPC `-32602` errors.

- Share protocol validation and error-result construction across stdio, embedded HTTP, and remote execution. Preserve existing known hidden tool names.
- Mark existing validation/rejection text as tool errors, preserve actionable validation and conflict messages, and redact typed internal/database/embedding and transport failure details from model-visible results.
- Validate unknown tools before HTTP scope classification; known tools still require their existing scopes, and group write guards remain in place.
- Keep `rpc_success` as the protocol outcome. Add nullable `tool_success` and `tool_error_kind` call-log columns with additive migrations and expose per-tool input/rejection/backend failure counts through admin statistics. Historical rows retain an unknown tool outcome. Expected input/rejection errors do not generate backend warning/health-error signals.
- Preserve `arguments: null` compatibility, plain-text REST validation/rejection messages, and failure markers for duplicate snapshots/branches, missing branch deletion, and empty apply selections. REST delegation propagates marked tool failures as HTTP errors.
- Limit partial-write warnings to tools that may mutate state; read-only failures do not imply a write.

MCP remains at `2024-11-05`; this does not depend on #236 or the notification fix.

## Review and validation

- Reviewed both dispatchers, remote response parsing, existing soft-error branches, HTTP admission, and the distinction between protocol and execution success.
- `cargo test -p memoria-mcp -p memoria-api --lib --offline`: 138 tests passed (104 API, 34 MCP).
- `cargo test -p memoria-mcp --test tools_unit --offline`: 9 tests passed.
- `cargo clippy -p memoria-mcp -p memoria-api --lib --offline -- -D warnings`: passed.
- `cargo test -p memoria-api --test api_e2e --no-run --offline`: passed; updated the unknown-tool assertions to expect `-32602`.
- Branch and snapshot end-to-end test targets compile successfully (`--no-run --offline`).
- Added regression coverage for null arguments, remote plain-text/JSON errors and HTML redaction, read-only error messaging, REST error propagation, and persisted per-tool error classification.
- `git diff --check`: passed.
- Database-backed end-to-end tests were not run locally because Docker is unavailable. Those tests remain for CI; local tests use lazy, unconnected pools and a local HTTP stub/transport failure to exercise protocol behavior.

## Follow-up CI fixes

- Update the snapshot quota integration test to expect the REST error response, and verify that rejection does not create an extra snapshot.
- Preserve legacy input-error text while carrying a typed input classification. Add database-independent regression coverage for the missing session-id message and its tool-error classification; the existing database integration assertions remain unchanged.

### CODE REVIEW

**Review by aptend — CHANGES_REQUESTED** [review]
Requesting changes for two error-classification issues:

1. memoria-mcp/src/tools.rs:820-824 and 930-934 stringify MemoriaError/sqlx::Error into new anyhow messages. This erases the source type that tool_result::execution_error relies on for redaction, so database details can be returned in isError content. Please preserve the underlying error chain, for example with anyhow::Error::from or context that retains the source.

2. memoria-mcp/src/tools.rs:427-431 similarly erases MemoriaError::InvalidTrustTier. An invalid trust_tier is consequently classified as a backend failure, gets a misleading partial-write warning, and pollutes backend health metrics even though execution has not begun. Please preserve the typed input error.

The relevant unit tests pass, but these paths are not covered by the current classification/redaction tests.

File: memoria/crates/memoria-mcp/src/remote.rs [hunk 1]
```diff
@@ -77,54 +79,91 @@ impl RemoteClient {
...
+        // Preserve actionable validation/conflict messages, without forwarding
+        // arbitrary HTML or proxy response bodies to the model.
+        let parsed: Value = serde_json::from_str(&body).unwrap_or(Value::Null);
+        let message = parsed
+            .get("error")
+            .and_then(Value::as_str)
+            .or_else(|| parsed.get("message").and_then(Value::as_str))
+            .or_else(|| (is_text && !body.trim().is_empty()).then_some(body.trim()))
+            .unwrap_or("Check the tool arguments and access permissions.");
+        let kind = match status.as_u16() {
```
Comment (aptend) [inline memoria/crates/memoria-mcp/src/remote.rs:223]: [P2] Ensure every remote REST response reaches this status mapper. `memory_branch_delete` (`remote.rs:818-824`) and `memory_rebuild_index` (`remote.rs:533-542`) currently discard the `reqwest::Response` and unconditionally return a successful MCP result. I reproduced the delete path against a local endpoint returning `409 Branch 'missing' not found`: `RemoteClient::call("memory_branch_delete", ...)` returned `{"content":[{"text":"Branch 'missing' deleted."}]}` with no `isError`. A 5xx from the rebuild endpoint is likewise reported as “requested.” This is the exact false-success behavior the PR is intended to eliminate, and clients/models can act on state changes that never happened. Please run both responses through `parse_response` (or `error_for_status` plus the same classification/redaction) before constructing success, with regression cases for the 409 and 5xx paths.

**Review by aptend — CHANGES_REQUESTED** [review]
Deep re-review at exact head `99280745e43307963a99ff0dcb8f48976e3d4dc0` (base/merge-base `99629991e8872b187e1b5831e997311d55cb64d4`): REQUEST_CHANGES.

The previous typed-error-chain findings are closed: `InvalidTrustTier`, `MemoriaError`, and `sqlx::Error` now retain their source types; classification/redaction walks the chain; and focused coverage confirms the behavior. One separate remote-dispatch blocker remains below.

Evidence on this head:
- `cargo test -p memoria-mcp -p memoria-api --lib --offline`: 145 passed.
- `cargo test -p memoria-mcp --test tools_unit --offline`: 9 passed.
- A temporary local-HTTP counterexample reproduced the remote branch-delete false success; the test was removed afterward.
- `git diff --check` passed; exact-head Check & Clippy, Unit Tests, and DB Tests are green.

File: memoria/crates/memoria-mcp/src/remote.rs [hunk 2]
```diff
@@ -77,54 +79,96 @@ impl RemoteClient {
...
+        crate::tool_result::error(e)
     }
 
     async fn parse_response(r: reqwest::Response) -> Result<Value> {
         let status = r.status();
         if status.is_success() {
-            return Ok(r.json().await?);
+            // 204 branch delete has no body. An empty success is not a JSON error.
+            let body = r.text().await?;
+            if body.trim().is_empty() {
```
Comment (aptend) [inline memoria/crates/memoria-mcp/src/remote.rs:170]: [P2] Scope empty-body success to endpoints that actually permit it. `parse_response` is shared by every remote tool, so returning `Value::Null` for any empty 2xx changes the previous fail-closed JSON behavior globally. Reproducer: make `POST /v1/memories` return `204 No Content`, then call `memory_store` with `{"content":"must be acknowledged"}`; this now returns a successful MCP result with text `Stored memory : ` and no `isError`, even though the required `memory_id`/`content` response is missing. The same pattern lets checkout/governance and other JSON-returning tools synthesize success from a malformed or truncated response. Keep empty 204 acceptance local to branch delete (or pass an explicit allow-empty contract), while JSON endpoints must still reject an empty body; add a negative regression for one JSON-returning tool.

**Review by aptend — CHANGES_REQUESTED** [review]
Deep re-review at exact head `976a336a8e6da8770597e4d5b7b3470f389d074f` (base/merge-base `99629991e8872b187e1b5831e997311d55cb64d4`): **REQUEST_CHANGES**.

The previous 409 branch-delete and 5xx rebuild false-success paths are now routed through `parse_response`, and their focused regressions pass. The latest change introduces one separate fail-open case: empty 2xx bodies are now accepted globally, including for tools whose successful contract requires JSON. The inline comment has the concrete counterexample and impact.

I re-read the complete review/comment/thread history, checked the incremental diff from `99280745e43307963a99ff0dcb8f48976e3d4dc0`, and revalidated the full PR diff. Local exact-head evidence: all 41 `memoria-mcp` library tests pass; a temporary local-HTTP regression test reproduces the empty-body false success and was removed afterward; `git diff --check` passes. Exact-head Check & Clippy and Unit Tests are green; DB Tests were still running at submission time.
